Fix bugs related to caching streamed volumes

aoldershaw opened the pull request:
<!--
Thanks for submitting a pull request to Concourse!

If you haven't already, feel free to [add yourself] as a contributor so that
you can add labels to your PR and re-trigger its builds if they fail.

Also check the [PR requirements] if you haven't already!
-->

[add yourself]: https://github.com/concourse/governance#individual-contributors
[PR requirements]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#pull-request-requirements

## Changes proposed by this PR

<!--
Summarize your changes as a checklist, leaving any unfinished work as unchecked
items. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->

with #6660, it was possible for a single worker to cache multiple copies
of the same resource_cache if it streamed in the same resource_cache
from multiple workers simultaneously.

this commit makes the worker_resource_caches unique on `(worker_name, resource_cache_id)`, (ignoring the `worker_base_resource_type_id` for purposes of uniqueness). in the event of a conflict, we indicate that the returned WorkerResourceCache is invalid, in which case we silently abort the InitializeResourceCache flow (much like how we silently abort when we have a unique code violation when updating the volume, which enforces there's only one volume for a single worker_resource_cache_id)

* [x] Prevent multiple `worker_resource_caches` from being created on a worker for a single `resource_cache`
* [x] Mark streamed resource cache volumes as unprivileged in Baggageclaim
* [x] Add a flag for disabling this feature. The feature is enabled by default since it *should* just be a transparent optimization, but with so much uncertainty around edge cases, it's safer to just provide an escape mechanism in case things go wrong

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

aoldershaw committed:
B: disallow duplicate resource caches on same worker

with #6660, it was possible for a single worker to cache multiple copies
of the same resource_cache if it streamed in the same resource_cache
from multiple workers simultaneously.

this commit makes the worker_resource_caches unique on `(worker_name,
resource_cache_id)`, (ignoring the `worker_base_resource_type_id` for
purposes of uniqueness). in the event of a conflict, we indicate that
the returned WorkerResourceCache is invalid, in which case we silently
abort the InitializeResourceCache flow (much like how we silently abort
when we have a unique code violation when updating the volume, which
enforces there's only one volume for a single worker_resource_cache_id)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
mark streamed cached volumes as unprivileged

we did the same thing in fetch_source for reasons that elude me, but it
probably makes sense to do the same thing here.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
add --disable-cache-streamed-volumes feature flag

it's enabled by default, but this is a way to opt-out

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

evanchaoli commented:
I wonder why you are adding a flag to disable streamed resource cache? Unless the feature has potential bugs, I cannot come out a reason to disable it.

aoldershaw commented:
> Unless the feature has potential bugs, I cannot come out a reason to disable it

I guess that's just it - we don't *think* it has any more bugs, but just today we already found a couple. There's a lot of edge cases and the worker package convoluted, so I figure it's safer to give people a way out if they encounter a serious bug after we release. It's on by default since it should just work fine, and we'll get rid of the flag when we know for sure it's stable

evanchaoli commented:
> it's safer to give people a way out if they encounter a serious bug after we release. It's on by default since it should just work fine, and we'll get rid of the flag when we know for sure it's stable

Make sense. In that case, do we also want a flag to turn off "get from local cache"? Maybe some user just want to always refresh resources.

evanchaoli commented:
With this PR,

https://github.com/concourse/concourse/blob/65933ae0a105ff3b7e6ad3027e1a62155cf728d9/atc/db/worker_resource_cache.go#L110-L132

may be updated accordingly, because given a worker name and a resource cache id, it should no longer find multiple resource caches.

aoldershaw committed:
S: simplify WorkerResourceCache.FindOrCreate

based on the fact that there can only be one worker_resource_cache for a
given worker + resource_cache pair

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented:
@evanchaoli thanks, good point - simplified the logic in b45990d3a4122e357b19b9d9ddb58590e3ac95c0.

> In that case, do we also want a flag to turn off "get from local cache"? Maybe some user just want to always refresh resources

That's a good question, I'm not too sure. Without the "get from local cache" logic you added, we still try to get from the cache  - the difference is we only check whether the *selected worker* has the resource cache, rather than checking all (compatible) workers. The approach of checking all compatible workers is seems to make more sense to me, with or without caching streamed volumes. WDYT @clarafu @muntac?

clarafu reviewed (commented):
Looks great to me :D great catches in those bugs!!

We talked about possibly adding an integration test for catching the privileged bug. One way that you thought of was to add an integration test that will hit baggageclaim but it might be better if we can think of a way for testing this in a concourse user perspective. Not sure if @muntac you have any thoughts about how we can go about doing that?

@aoldershaw, I think we might want to turn off the "get from local cache" too if the user disables caching streamed volumes. This is because if we don't then we will only ever fetch the volume on one worker and then everytime we will just find that one fetched volume and it will result in a lot of streaming (which is pretty expensive). I think we want the volume to be spread across all workers so that we don't need to do a lot of streaming.

aoldershaw commented on atc/db/volume.go:
```diff
@@ -371,13 +371,20 @@ func (volume *createdVolume) initializeResourceCache(resourceCache UsedResourceC
 	if err != nil {
 		return err
 	}
+	if !valid {
```
Good point. To be honest, I'm not sure exactly what setting `team_id` to `nil` does, but it's probably safest for us to do that here as well I suppose

aoldershaw replied:
Uncommenting the `Set("team_id", nil)` on the original update doesn't seem to cause any tests to fail - I wonder if it's just legacy logic or if it's still important but untested 🤔

aoldershaw replied:
Actually, thinking about it more, I think it probably doesn't make sense to set `team_id` to `nil` - I suspect it was only set to `nil` since even if a resource cache volume was produced on a workload for team X, it can still be used by a workload for team Y. But since this case is for when the volume will *not* become the resource cache, we shouldn't need to do this.

In the event of a unique violation below: https://github.com/concourse/concourse/blob/b45990d3a4122e357b19b9d9ddb58590e3ac95c0/atc/db/volume.go#L389-L403

...we still don't update `team_id` to `nil`, and this really represents a similar case - another volume "won the race" to being the resource cache.

clarafu replied:
Ohhh yeah I see, this should just not work if there was a dup anyways because of the unique index on `worker_resource_cache_id`. Cool thanks for digging into it for me :D

aoldershaw committed:
make ypath helpers ensure no errors

to avoid clutter in the tests and for consistency with the rest of the
internal helpers (e.g. cmdtest)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu commented:
@aoldershaw I think the rest is good and I could approve it after we turn off the "get from local cache" if the flag is disables caching streamed volumes. What do you think about my comment on why I think that we should do that, do you think differently or maybe I'm missing something?

aoldershaw commented:
@clarafu ah whoops, I missed that part of your comment somehow 😁 sure, seems reasonable. Also working on some more integration tests

aoldershaw committed:
generate dynamic docker-compose.yml for integration

this way, we don't need to worry about the main docker-compose.yml
drifting from the override as much, since the override contains a
complete copy of the worker service. this way, we just clone the
existing service and modify it as needed in code.

also enhances the assertion to ensure we pass a COW volume in as both
the base image and the input

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented:
@clarafu spent some time trying to write an integration test for the "mark as unprivileged" case, but don't think it's so easily testable. I found https://github.com/concourse/concourse/commit/0ee25ca23253d9fabc0fab48f4c1bbf774989492, which suggests marking the volume as unprivileged "optimizes for the most common case of using images or inputs for an unprivileged container" - so, it's not a bug per se in that it affects outward behaviour, but would be a bit slower for unprivileged containers.

aoldershaw committed:
B: only get from local cache when caching enabled

otherwise, we'll always only run the get on a single worker, meaning
we'll need to perform a lot more streaming

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
add missing error check

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu reviewed (approved):
LGTM! 👍  

We already shipped the migration for the original caching streamed volumes PR to our `ci` cluster, I guess we just have to remember to edit the unique index in `ci` to the changes you made here?

aoldershaw commented:
@clarafu yeah, think maybe i’ll run the down migration and then redeploy (that way, we can verify the down migration as well)
